src/player.spec.ts

```diff
@@ -75,4 +76,89 @@ describe("Player", () => {
       expect(orderedEvents).to.be.equal("1c,1o,1k,1t");
     });
   });
+
+  describe("canFullyChargeDuringIncomePhase", () => {
+    const tests: {
+      name: string;
+      power: Power;
+      brainstone: BrainstoneArea;
+      events: Event[];
+      expected: boolean;
+    }[] = [
+      {
+        name: "no events - not fully charged",
+        power: new Power(1),
+        brainstone: null,
+        events: [],
+        expected: false,
+      },
+      {
+        name: "no events - not fully charged (brainstone)",
+        power: new Power(),
+        brainstone: BrainstoneArea.Area1,
+        events: [],
+        expected: false,
+      },
+      {
+        name: "no events - fully charged",
+        power: new Power(0, 0, 1),
+        brainstone: null,
+        events: [],
+        expected: true,
+      },
+      {
+        name: "no events - fully charged (brainstone)",
+        power: new Power(),
+        brainstone: BrainstoneArea.Area3,
+        events: [],
+        expected: true,
+      },
+      {
+        name: "no events - no tokens",
+        power: new Power(),
+        brainstone: null,
+        events: [],
+        expected: true,
+      },
+      {
+        name: "events - will charge fully",
+        power: new Power(0, 1, 0),
+        brainstone: null,
+        events: Event.parse(["+3pw", "+1t"], Command.ChooseIncome),
+        expected: true,
+      },
+      {
+        name: "events - will charge fully (brainstone)",
+        power: new Power(),
+        brainstone: BrainstoneArea.Area2,
+        events: Event.parse(["+3pw", "+1t"], Command.ChooseIncome),
+        expected: true,
+      },
+      {
+        name: "events - will not charge fully",
+        power: new Power(0, 1, 0),
+        brainstone: null,
+        events: Event.parse(["+2pw", "+1t"], Command.ChooseIncome),
+        expected: false,
+      },
+      {
+        name: "events - will not charge fully (brainstone)",
+        power: new Power(),
+        brainstone: BrainstoneArea.Area2,
+        events: Event.parse(["+2pw", "+1t"], Command.ChooseIncome),
+        expected: false,
+      },
+    ];
+
+    tests.forEach((test) => {
```

Comment:
I prefer `for (const test of tests) {` unless there's a compelling reason not to do it that way. It's more natural / works with async.